File: cigarsTest.py
#!/usr/bin/env python

#Copyright (C) 2006-2012 by Benedict Paten (benedictpaten@gmail.com)
#
#Released under the MIT license, see LICENSE.txt

import unittest

import os
import sys
import random

from .bioio import getTempFile

from .bioio import cigarRead
from .bioio import cigarWrite
from .bioio import getRandomPairwiseAlignment
from .bioio import system
from sonLib.bioio import TestStatus
from sonLib.bioio import logger

class TestCase(unittest.TestCase):
    
    def setUp(self):
        self.testNo = TestStatus.getTestSetup()
        self.tempFiles = []
        unittest.TestCase.setUp(self)
    
    def tearDown(self):
        for tempFile in self.tempFiles:
            os.remove(tempFile)
        unittest.TestCase.tearDown(self)
    
    def testCigarReadWrite(self):
        """Tests the C code for reading and writing cigars against the python parser for cigars.
        """
        tempFile = getTempFile()
        self.tempFiles.append(tempFile)
        for test in range(0, self.testNo):
            pairwiseAlignmentNumber = random.choice(range(10))
            l = [ getRandomPairwiseAlignment() for i in range(pairwiseAlignmentNumber) ]
            fileHandle = open(tempFile, 'w')
            
            keepProbs = random.random() > 0.5
            if keepProbs == False:
                for pA in l:
                    for op in pA.operationList:
                        op.score = 0.0
            
            for pairwiseAlignment in l:
                cigarWrite(fileHandle, pairwiseAlignment, keepProbs)
            fileHandle.close()

            #Now call sonLib_cigarsTest and read and write chains
            command = "sonLib_cigarTest %s %s" % (tempFile, keepProbs)
            #return
            system(command)
            
            #Now check the chain is okay
            fileHandle = open(tempFile, 'r')
            l.reverse()
            
            for pairwiseAlignment in cigarRead(fileHandle):
                pairwiseAlignment2 = l.pop()
                cigarWrite(sys.stdout, pairwiseAlignment, keepProbs)
                cigarWrite(sys.stdout, pairwiseAlignment2, keepProbs)
                assert pairwiseAlignment == pairwiseAlignment2
            assert len(l) == 0
            fileHandle.close()
        
if __name__ == '__main__':
    unittest.main()

File: C/tests/cigarsTest.c
/*
 * Copyright (C) 2006-2012 by Benedict Paten (benedictpaten@gmail.com)
 *
 * Released under the MIT license, see LICENSE.txt
 */

#include <stdlib.h>
#include <limits.h>
#include <assert.h>
#include <string.h>

#include "fastCMaths.h"
#include "avl.h"
#include "commonC.h"
#include "bioioC.h"
#include "pairwiseAlignment.h"

int main(int argc, char *argv[]) {
    int32_t i;
    struct List *pAs;
    FILE *fileHandle;
    struct PairwiseAlignment *pA;
    int32_t keepProbs;

    assert(argc == 3);
    if(strcmp(argv[2], "True") == 0) {
        keepProbs = TRUE;
    }
    else {
        assert(strcmp(argv[2], "False") == 0);
        keepProbs = FALSE;
    }

    pAs = constructEmptyList(0, (void (*)(void *))destructPairwiseAlignment);
    fileHandle = fopen(argv[1], "r");
    pA = cigarRead(fileHandle);
    while(pA != NULL) {
        listAppend(pAs, pA);
        pA = cigarRead(fileHandle);
    }
    fclose(fileHandle);

    fileHandle = fopen(argv[1], "w");
    for(i=0; i<pAs->length; i++) {
        cigarWrite(fileHandle, pAs->list[i], keepProbs);
    }
    fclose(fileHandle);

    return 0;
}
